From 92dd40adcc53ce85eeae259b0eda35b77846f3c5 Mon Sep 17 00:00:00 2001
From: Jashan Kamboj <jashank42@gmail.com>
Date: Wed, 22 Jul 2015 01:06:50 +0530
Subject: [PATCH 14/31] add link check_access

Signed-off-by: Jashan Kamboj <jashank42@gmail.com>
---
 src/mds/Server.cc | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/mds/Server.cc b/src/mds/Server.cc
index 582f0fd..033909f 100644
--- a/src/mds/Server.cc
+++ b/src/mds/Server.cc
@@ -4677,6 +4677,12 @@ void Server::handle_client_link(MDRequestRef& mdr)
   if (!mds->locker->acquire_locks(mdr, rdlocks, wrlocks, xlocks))
     return;
 
+  if (!check_access(mdr, targeti, MAY_WRITE))
+    return;
+
+  if (!check_access(mdr, dir->get_inode(), MAY_WRITE))
+    return;
+
   // go!
   assert(g_conf->mds_kill_link_at != 1);
 
-- 
1.9.1

