From 230eacce5c3e274d5adf9e3ca518df0c00173d88 Mon Sep 17 00:00:00 2001
From: Jashan Kamboj <jashank42@gmail.com>
Date: Tue, 18 Aug 2015 20:39:58 +0530
Subject: [PATCH 30/31] test/libcephfs/access: add update_after_unlink
 test

Signed-off-by: Jashan Kamboj <jashank42@gmail.com>
---
 src/test/libcephfs/access.cc | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/src/test/libcephfs/access.cc b/src/test/libcephfs/access.cc
index c6c797d..4f16d1d 100644
--- a/src/test/libcephfs/access.cc
+++ b/src/test/libcephfs/access.cc
@@ -202,6 +202,25 @@ TEST(AccessTest, ReadOnly){
   ceph_shutdown(admin);
 }
 
+TEST(AccessTest, update_after_unlink){
+  struct ceph_mount_info *admin;
+  ASSERT_EQ(0, ceph_create(&admin, NULL));
+  ASSERT_EQ(0, ceph_conf_parse_env(admin, NULL));
+  ASSERT_EQ(0, ceph_conf_read_file(admin, NULL));
+  ASSERT_EQ(0, ceph_mount(admin, "/"));
+  ASSERT_EQ(0, ceph_mkdir(admin, "/foo", 0755));
+  int fd = ceph_open(admin, "/foo/bar", O_CREAT|O_WRONLY, 0755);
+  ceph_unlink(admin, "/foo/bar");
+  ASSERT_GE(ceph_write(admin, fd, "foo", 3, 0), 0);
+  ASSERT_GE(ceph_fchmod(admin, fd, 0777), 0);
+  ASSERT_GE(ceph_ftruncate(admin, fd, 0), 0);
+  ASSERT_GE(ceph_fsetxattr(admin, fd, "user.any", "bar", 3, 0), 0);
+  ceph_close(admin,fd);
+  
+  ASSERT_EQ(0, ceph_rmdir(admin, "/foo"));
+  ceph_shutdown(admin);
+}
+
 int main(int argc, char **argv)
 {
   ::testing::InitGoogleTest(&argc, argv);
-- 
1.9.1

